ingest/src/main/java/com/microsoft/azure/kusto/ingest/ExponentialRetry.java

```diff
@@ -19,30 +27,36 @@ public ExponentialRetry(ExponentialRetry other) {
         this.maxJitterSecs = other.maxJitterSecs;
     }
 
-    private double getCurrentSleepSecs() {
-        return sleepBaseSecs * (float) Math.pow(2, currentAttempt);
-    }
-
-    public int getCurrentAttempt() {
-        return currentAttempt;
-    }
-
-    public String nextTimeToSleep() {
-        return String.format("%s +~ %s seconds", getCurrentSleepSecs(), maxJitterSecs);
-    }
-
-    public boolean doBackoff() {
-        double jitterSecs = (float) Math.random() * maxJitterSecs;
-        double sleepMs = (getCurrentSleepSecs() + jitterSecs) * 1000;
-        try {
-            Thread.sleep((long) sleepMs);
-        } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            throw new RuntimeException("Interrupted while sleeping", e);
+    public <T> T execute(KustoCheckedFunction<Integer, T> function) throws IngestionClientException, IngestionServiceException {
+        for (int currentAttempt = 0; currentAttempt < maxAttempts; currentAttempt++) {
+            log.info("Attempt {}", currentAttempt);
+
+            try {
+                T result = function.apply(currentAttempt);
+                if (result != null) {
+                    return result;
+                }
+            }
+            catch (Exception e) {
+                log.error("Error is permanent, stopping", e);
+                throw e;
+            }
+
+            double currentSleepSecs = sleepBaseSecs * (float) Math.pow(2, currentAttempt);
+
+            log.info("Operation failed, trying again after sleep of %{} +~ %{} seconds", currentSleepSecs, maxJitterSecs);
```

Comment:
Why not move this to after where `sleepMs` is defined, and change to:
`log.info("ExponentialRetry.execute: Attempt {} failed, trying again after sleep of {} milliseconds", currentAttempt, sleepMs);`

Reply:
Great idea (I used seconds)

Also, we don't need to write the class name, that's part of the log:
[main] INFO com.microsoft.azure.kusto.ingest.ExponentialRetry - execute: Attempt 0